modules/v2x/v2x_proxy/app/utils.cc

```diff
@@ -0,0 +1,600 @@
+  sim_oslight_ = std::make_shared<OSLight>();
+  obu_light_ = std::make_shared<ObuLight>();
+  this->reset();
+}
+
+InternalData::~InternalData() {
+  delete[] remaining_time_;
+  delete[] msg_timestamp_;
+}
+
+void InternalData::reset() {
+  oslight_ = nullptr;
+  sim_oslight_->Clear();
+  sim_sending_count_ = 0;
+  intersection_id_ = -1;
+  change_color_timestamp_ = 0.0;
+}
+
+bool InternalData::TrafficLightProc(
+    const std::shared_ptr<::apollo::hdmap::HDMap> &hdmap, double distance,
+    ::apollo::v2x::RoadTrafficLight *msg,
+    ::apollo::v2x::RoadTrafficLight *sim_msg) {
+  if (nullptr == msg) {
+    return false;
+  }
+  if (!msg->has_gps_x_m() || !msg->has_gps_y_m()) {
+    AERROR << "Error::v2x traffic_light ignore, gps point is null";
+    return false;
+  }
+  ::apollo::common::PointENU point;
+  point.set_x(msg->gps_x_m());
+  point.set_y(msg->gps_y_m());
+  if (0 == msg->single_traffic_light_size()) {
+    AERROR << "Error::v2x traffic_light ignore, size of single light is 0.";
+    return false;
+  }
+  if (0 == msg->single_traffic_light(0).traffic_light_type_size()) {
+    return false;
+  }
+  ::apollo::hdmap::LaneInfoConstPtr laneinfo;
+  double dummy_s = 0, dummy_l = 0;
+  if (0 != hdmap->GetNearestLane(point, &laneinfo, &dummy_s, &dummy_l)) {
+    return false;
+  }
+  std::vector<::apollo::hdmap::SignalInfoConstPtr> signals;
+  if (0 != hdmap->GetForwardNearestSignalsOnLane(point, distance, &signals)) {
+    AERROR << "Error::v2x traffic_light ignore, hdmap get no signals."
+           << "traffic light size : " << signals.size() << " "
+           << std::setiosflags(std::ios::fixed) << std::setprecision(11)
+           << "Point:x=" << point.x() << ",y=" << point.y();
+    return false;
+  }
+  if (signals.empty()) {
+    AERROR << "Get traffic light size : " << signals.size() << " "
+           << std::setiosflags(std::ios::fixed) << std::setprecision(11)
+           << "Point:x=" << point.x() << ",y=" << point.y();
+    return false;
+  }
+  AINFO << "the size of traffic light from HDMap is: " << signals.size();
+  auto tl_type = msg->single_traffic_light(0).traffic_light_type(0);
+  auto color = msg->single_traffic_light(0).color();
+  auto remaining_time = msg->single_traffic_light(0).color_remaining_time_s();
+  auto next_color = msg->single_traffic_light(0).next_color();
+  auto next_remaining_time =
+      msg->single_traffic_light(0).next_remaining_time_s();
+  msg->clear_single_traffic_light();
+  for (size_t i = 0; i < signals.size(); i++) {
+    auto signal_info = signals[i];
+    auto single = msg->add_single_traffic_light();
+    single->set_id(signal_info->id().id());
+    single->add_traffic_light_type(tl_type);
+    single->set_color(color);
+    single->set_color_remaining_time_s(remaining_time);
+    single->set_next_color(next_color);
+    single->set_next_remaining_time_s(next_remaining_time);
+  }
+  return true;
+}
+
+bool InternalData::GenerateSimMsg() {
+  if (nullptr == oslight_) {
```

Comment:
This function is too long.

Reply:
This function will be rewritten in the next version. Prior to this, repeated tests were required on the main line.